model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/initializer/OfflinePersistentUserSessionLoader.java

```diff
@@ -53,46 +54,67 @@ public OfflinePersistentUserSessionLoader(int sessionsPerSegment) {
 
     @Override
     public void init(KeycloakSession session) {
-        UserSessionPersisterProvider persister = session.getProvider(UserSessionPersisterProvider.class);
+    }
 
-        // TODO: check if update of timestamps in persister can be skipped entirely
-        int clusterStartupTime = session.getProvider(ClusterProvider.class).getClusterStartupTime();
 
-        log.debugf("Clearing detached sessions from persistent storage and updating timestamps to %d", clusterStartupTime);
+    @Override
+    public OfflinePersistentLoaderContext computeLoaderContext(KeycloakSession session) {
+        UserSessionPersisterProvider persister = session.getProvider(UserSessionPersisterProvider.class);
+        int sessionsCount = persister.getUserSessionsCount(true);
 
-        persister.clearDetachedUserSessions();
-        persister.updateAllTimestamps(clusterStartupTime);
+        return new OfflinePersistentLoaderContext(sessionsCount, sessionsPerSegment);
     }
 
 
     @Override
-    public OfflinePersistentUserSessionLoaderContext computeLoaderContext(KeycloakSession session) {
-        UserSessionPersisterProvider persister = session.getProvider(UserSessionPersisterProvider.class);
-        int sessionsCount = persister.getUserSessionsCount(true);
+    public OfflinePersistentWorkerContext computeWorkerContext(OfflinePersistentLoaderContext loaderCtx, int segment, int workerId, List<OfflinePersistentWorkerResult> previousResults) {
```

Comment:
Is `previousResults` used at all? IMHO per code, it will be always empty (which would be correct)

Now that the `lastSessionRefresh`/`lastSessionId` is used in `JpaUserSessionPersisterProvider.loadUserSessions`, if `previousResults` and `OfflinePersistentWorkerContext` was updated on-the-fly, then setting `firstResult` in that method would not be valid. Setting it while changing `previousResults` and thus `lastSession...` properties, some of the stored sessions could've been skipped.